From dc48c11af444283280ff4063e4be825ed26f0f55 Mon Sep 17 00:00:00 2001
From: Ahmad Samir <a.samirh78@gmail.com>
Date: Mon, 13 Oct 2025 08:23:37 +0300
Subject: [PATCH 450/553] tst_QSettings: split the testEscapes() test:
 testunEscapedStringList

Easier to debug.

Use example.org instead of software.org, the latter is regitered and we
shouldn't use it in tests.

Pick-to: 6.8 6.5
Change-Id: Iee15c510a3c225f0142a518b7ba75c68a7bb0fa2
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
Reviewed-by: Thiago Macieira <thiago.macieira@intel.com>
(cherry picked from commit 0563d7831c87f1fc6336652e2a416635c09bf1ff)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../corelib/io/qsettings/tst_qsettings.cpp    | 101 ++++++++++++------
 1 file changed, 69 insertions(+), 32 deletions(-)

diff --git a/tests/auto/corelib/io/qsettings/tst_qsettings.cpp b/tests/auto/corelib/io/qsettings/tst_qsettings.cpp
index 46a89c5631c..bb666b98612 100644
--- a/tests/auto/corelib/io/qsettings/tst_qsettings.cpp
+++ b/tests/auto/corelib/io/qsettings/tst_qsettings.cpp
@@ -190,6 +190,8 @@ private slots:
     void testUnescapedKeys();
     void testEscapedStringList_data();
     void testEscapedStringList();
+    void testUnescapedStringList_data();
+    void testUnescapedStringList();
     void testNormalizedKey_data();
     void testNormalizedKey();
     void testVariantTypes_data() { populateWithFormats(); }
@@ -2836,17 +2838,6 @@ void tst_QSettings::testEscapes()
 {
     QSettings settings(QSettings::UserScope, "software.org", "KillerAPP");
 
-#define testUnescapedStringList(escStrList, plainStrList, reescStrList) \
-    { \
-        QStringList plainList(plainStrList); \
-        QByteArray escList(escStrList); \
-        QByteArray reescList(reescStrList); \
-        QCOMPARE(iniUnescapedStringList(escList), plainList); \
-        QCOMPARE(iniEscapedStringList(plainList), reescList); \
-        QCOMPARE(iniUnescapedStringList(reescList), plainList); \
-    } \
-
-
 #define testVariant(val, escStr, func) \
     { \
         QVariant v(val); \
@@ -2862,27 +2853,6 @@ void tst_QSettings::testEscapes()
         QCOMPARE(v.toString(), QString(vStr)); \
     }
 
-    /*
-      Test .ini syntax that cannot be generated by QSettings (but can be entered by users).
-    */
-    testUnescapedStringList("", "", "");
-    testUnescapedStringList("\"\"", "", "");
-    testUnescapedStringList("\"abcdef\"", "abcdef", "abcdef");
-    testUnescapedStringList("\"\\?\\'\\\"\"", "?'\"", "?'\\\"");
-    testUnescapedStringList("\\0\\00\\000\\0000000\\1\\111\\11111\\x\\x0\\xABCDEFGH\\x0123456\\",
-                            QString() + QChar(0) + QChar(0) + QChar(0) + QChar(0) + QChar(1)
-                            + QChar(0111) + QChar(011111) + QChar(0) + QChar(0xCDEF) + "GH"
-                            + QChar(0x3456),
-                            "\\0\\0\\0\\0\\x1I\xE1\x89\x89\\0\xEC\xB7\xAFGH\xE3\x91\x96");
-    testUnescapedStringList(QByteArray("\\c\\d\\e\\f\\g\\$\\*\\\0", 16), "\f", "\\f");
-    testUnescapedStringList("\"a\",  \t\"bc \", \"  d\" , \"ef  \" ,,g,   hi  i,,, ,",
-                            QStringList() << "a" << "bc " << "  d" << "ef  " << "" << "g" << "hi  i"
-                                          << "" << "" << "" << "",
-                            "a, \"bc \", \"  d\", \"ef  \", , g, hi  i, , , , ");
-    testUnescapedStringList("a ,  b   ,   c   d   , efg   ",
-                            QStringList() << "a" << "b" << "c   d" << "efg",
-                            "a, b, c   d, efg");
-
     // streaming qvariant into a string
     testVariant(QString("Hello World!"), QString("Hello World!"), toString);
     testVariant(QString("Hello, World!"), QString("Hello, World!"), toString);
@@ -3023,6 +2993,73 @@ void tst_QSettings::testEscapedStringList()
     QCOMPARE(iniUnescapedStringList(escapedList), plainStrList);
 }
 
+void tst_QSettings::testUnescapedStringList_data()
+{
+    QTest::addColumn<QByteArray>("escStrList");
+    QTest::addColumn<QStringList>("plainStrList");
+    QTest::addColumn<QByteArray>("reescStrList");
+
+    /*
+      Test .ini syntax that cannot be generated by QSettings (but can be entered by users).
+    */
+    QTest::newRow("empty")
+        << ""_ba
+        << QStringList{u""_s}
+        << ""_ba;
+
+    QTest::newRow("empty-double-quotes")
+        << "\"\""_ba
+        << QStringList{u""_s}
+        << ""_ba;
+
+    QTest::newRow("plain-quoted-string")
+        << "\"abcdef\""_ba
+        << QStringList{u"abcdef"_s}
+        << "abcdef"_ba;
+
+    QTest::newRow("backslash-non-letter-characters")
+        << "\"\\?\\'\\\"\""_ba
+        << QStringList{u"?'\""_s}
+        << "?'\\\""_ba;
+
+    const std::array arr = {QChar(0), QChar(0), QChar(0), QChar(0), QChar(1),
+                            QChar(0111), QChar(011111), QChar(0), QChar(0xCDEF),
+                            QChar(u'G'), QChar(u'H'), QChar(0x3456)};
+    QTest::newRow("array-of-qchar")
+        << "\\0\\00\\000\\0000000\\1\\111\\11111\\x\\x0\\xABCDEFGH\\x0123456\\"_ba
+        << QStringList{QString{arr}}
+        << "\\0\\0\\0\\0\\x1I\xE1\x89\x89\\0\xEC\xB7\xAFGH\xE3\x91\x96"_ba;
+
+    QTest::newRow("backslash-escapes")
+        << QByteArray("\\c\\d\\e\\f\\g\\$\\*\\\0", 16)
+        << QStringList{u"\f"_s}
+        << "\\f"_ba;
+
+    QTest::newRow("double-quotes-tab-character")
+        << "\"a\",  \t\"bc \", \"  d\" , \"ef  \" ,,g,   hi  i,,, ,"_ba
+        << QStringList{u"a"_s, u"bc "_s, u"  d"_s, u"ef  "_s, u""_s, u"g"_s,
+                       u"hi  i"_s, u""_s, u""_s, u""_s, u""_s}
+        << "a, \"bc \", \"  d\", \"ef  \", , g, hi  i, , , , "_ba;
+
+    QTest::newRow("abcdefg-extra-whitespaces")
+        << "a ,  b   ,   c   d   , efg   "_ba
+        << QStringList{u"a"_s, u"b"_s, u"c   d"_s, u"efg"_s}
+        << "a, b, c   d, efg"_ba;
+}
+
+void tst_QSettings::testUnescapedStringList()
+{
+    QFETCH(QByteArray, escStrList);
+    QFETCH(QStringList, plainStrList);
+    QFETCH(QByteArray, reescStrList);
+
+    QSettings settings(QSettings::UserScope, "example.org", "KillerAPP");
+
+    QCOMPARE(iniUnescapedStringList(escStrList), plainStrList);
+    QCOMPARE(iniEscapedStringList(plainStrList), reescStrList);
+    QCOMPARE(iniUnescapedStringList(reescStrList), plainStrList);
+}
+
 #endif
 
 void tst_QSettings::testCaseSensitivity()
-- 
2.51.2

